충남대 BE_윤정훈 2주차 과제 (3단계)

yunjunghun0116 opened the pull request:
### 2단계 과제 리팩토링 과정
1. 2단계 코드리뷰 바탕으로 코드 리팩토링 수행
2. Interceptor 에서 Header에 대한 검증 로직을 수행
3. Interceptor 기반 테스트 로직 수행
4. 상품의 카카오 포함 검증 로직을 Controller 계층에서 Service 계층으로 분리
5. 인증 로직과 멤버 로직의 분리

### 3단계 과제 수행 과정
1. README.md 작성
2. 위시 리스트에 대한 프로세스와 계층 구상
3. 추가, 수정 관련 DTO 및 도메인 생성
4. 계층 별로 로직을 작성
5. 테스트 코드 작성

### 고민을 가장 많이 한 부분 및 느낀점
가독성 좋고 잘 짠 코드를 짜기 위해서 어떻게 코드를 짜는게 좋을까?
1. 함수명만으로 어떤 역할을 수행하는지 알도록
2. 변수명을 통해 어떤 정보를 갖고있는 객체인지 알도록
3. 다른 계층과의 의존은 줄일 수 있으면 줄이자
라는 스스로의 규칙? 안에서 최대한 성능이 좋은 코드를 작성하도록 했습니다.

위 3가지 규칙에 위배되는 코드나 혹시 추가적으로 개선 가능한 코드는 알려주시면 최대한 반영하겠습니다!

yunjunghun0116 committed:
docs: 2주차 위시리스트 README.md 작성 및 0단계 수행

yunjunghun0116 committed:
move: 상품 관리의 코드를 가져온다

상품관리의 코드를 가져온다.

yunjunghun0116 committed:
docs: README 작성

README.md 기능 요구 사항 작성 및 코드 옮겨오기 수행(완)

yunjunghun0116 committed:
cherry-pick: README.md 커밋 가져오기

README.md 커밋 가져오기

yunjunghun0116 committed:
refactor: repository의 객체 생성 역할을 service계층으로 이동하여 각 계층의 역할을 완전 분리

계층의 역할을 완전 분리하기 위한 refactor 작업 수행

yunjunghun0116 committed:
refactor: ProductOption에서의 Repository와 Service계층간 역할 분리

역할 분리작업 수행

yunjunghun0116 committed:
refactor: 와일드카드 임포트를 허용하지 않는 룰 적용

와일드카드 임포트 허용하지 않도록 import 수정 및 intellij 옵션 수정

yunjunghun0116 committed:
feat: PriceLessThanZeroException 예외 처리 기능 추가

상품의 가격이나, 상품옵션의 추가금액이 0원보다 작을 수 없음

yunjunghun0116 committed:
feat: 존재하지 않는 객체에 대한 접근 예외 처리

존재하지 않는 객체에 대한 접근 예외 처리

yunjunghun0116 committed:
feat: id를 통한 접근시 객체를 찾지 못할경우 NotFoundElementException 예외 발생

잘못된 접근을 통한 리소스 조회시 NotFound예외 발생

yunjunghun0116 committed:
feat: Controller계층에서 상품 이름 길이에 대한 검증 수행

Controller계층에서 상품이름 길이에 대한 검증을 수행할 수 있도록 RequestDto에 validation을 추가

yunjunghun0116 committed:
feat: validation 을 사용해서 입력을 컨트롤러 계층에서 검증 작업 수행

컨트롤러 계층에서 입력값에 대한 검증을 수행

yunjunghun0116 committed:
test: 입력 검증을 Controller 계층에서 해줌에 따라 Controller계층의 테스트 추가

입력 검증 방식의 변경에 따라 계층 테스트 추가

yunjunghun0116 committed:
test: 다양한 경우에 따른 입력 검증을 수행하는 테스트 추가

다양한 경우에 따른 입력검증을 수행하는 테스트 추가

yunjunghun0116 committed:
feat: 정규식을 적용해서 카카오 포함된 문구 확인

정규식을 사용
정규식 : (?!.*카카오).* 를 사용해서 맨 앞에서부터 문자를 탐색하며 카카오를 찾고, 만약 카카오가 찾아지면 안된다.
뒤에 .*이 붙은 이유는 카카오를 제외한 어떤 문자든 허용하기 위해서

yunjunghun0116 committed:
feat: 상품 이름의 최소 길이 지정

상품 이름의 최소길이를 지정해주어 빈 이름이 넘어오지 않도록 설정

yunjunghun0116 committed:
test: 빈 이름을 가진 오류상품 생성하는 테스트 추가

빈 이름을 가진 오류상품 생성요청시 예외 발생

yunjunghun0116 committed:
test: 공백을 포함한 정상 상품 생성하는 테스트 생성

공백을 포함한 정상 상품을 생성하는 테스트 작성

yunjunghun0116 committed:
refactor: 계층간 역할분리를 위한 레포지토리에서의 역할 분리

레포지토리에서의 역할 분리

yunjunghun0116 committed:
refactor: 나만의 RULE에 맞게 DB와 엔티티 관련 작업을 Service -> Repository로 역할을 변경

Repository의 역할을 변경, Service에서는 엔티티 <-> DTO, Repository 에서는 엔티티 <-> DB 데이터

yunjunghun0116 committed:
feat: imageUrl, 상품옵션의 필드 등에 추가적인 입력 검증 조건 추가

입력 검증 추가

yunjunghun0116 committed:
test: 상품 옵션 입력 검증 테스트 수행

상품 옵션 입력 검증 테스트 수행

yunjunghun0116 committed:
Merge branch 'yunjunghun0116' into step1

yunjunghun0116 committed:
remove: 사용하지 않는 예외 제거

사용하지 않는 예외 제거

yunjunghun0116 committed:
refactor: newline at end of file 스타일 가이드 준수

newline at end of file 이라는 구글의 스타일 가이드를 준수하기 위해 개행을 추가

yunjunghun0116 committed:
refactor: Entity -> ResponseDto로 변환하는 작업은 Service계층에서

Entity -> ResponseDto로 변환하는 작업은 Service계층에서

yunjunghun0116 committed:
refactor: Entity -> ResponseDto로 변환하는 작업은 Service계층에서(상품옵션)

상품옵션도 Entity -> ResponseDto로 변환하는 작업은 Service계층에서

yunjunghun0116 committed:
refactor: 상품 save 메서드의 역할 분리

상품의 save메서드 역할 분리

yunjunghun0116 committed:
refactor: 상품옵션 save 메서드의 역할 분리

상품옵션 save메서드의 역할 분리

yunjunghun0116 committed:
refactor: 상품 수정 메서드의 역할 분리 -> 업데이트 따로, Entity to Dto 변환 따로

상품 수정 메서드의 역할을 분리
업데이트와 엔티티와 dto간 변환을 분리

yunjunghun0116 committed:
refactor: 상품 옵션 수정 메서드의 역할 분리 -> 업데이트 따로, Entity to Dto 변환 따로

상품 옵션 수정 메서드의 역할을 분리
업데이트를 따로,
엔티티를 response dto로 변환하는것도 따로 역할분리 작업을 수행

yunjunghun0116 committed:
refactor: option Row Mapper 코드 분리 작업 수행

option Row Mapper 코드 분리 작업을 수행

yunjunghun0116 committed:
refactor: 상품 row Mapper 코드 분리 작업 수행

row Mapper 상품 코드 분리 작업 수행

yunjunghun0116 committed:
feat: 존재하지 않는 상품id에 대한 옵션 생성시 외래키 제약조건 위배 예외를 발생

외래키 제약조건 위배 예외를 발생

yunjunghun0116 committed:
feat: 외래키 제약조건에 위배되는 경우에 대한 예외 처리

외래키 제약 조건에 위배되는 경우 ForeignKeyConstraintViolationException 예외를 발생

yunjunghun0116 committed:
feat: 존재하지 않는 상품에 대해 상품옵션을 생성하려고 하면 외래키제약조건위배예외 발생

존재하지 않는 상품에 대해 상품옵션을 생성하려고 하면 외래키제약조건위배예외 발생

yunjunghun0116 committed:
test: 존재하지 않는 상품에 대한 옵션 생성하는 테스트 추가

존재하지 않는 상품에 대한 옵션을 생성할경우 409 에러 Conflict 발생

yunjunghun0116 committed:
Merge pull request #1 from kakao-tech-campus-2nd-step2/yunjunghun0116

충남대 BE_윤정훈 2주차 과제(1단계) (#108)

yunjunghun0116 committed:
refactor: 객체지향 생활체조원칙에 의한 코드 개선

객체지향 생활체조원칙에 의한 코드 개선
각 지역변수 var 통일

yunjunghun0116 committed:
refactor: 상품과 상품요청DTO간 의존관계 제거

상품과 상품요청DTO간 의존관계 제거

yunjunghun0116 committed:
refactor: 옵션과 옵션요청DTO간 의존관계 제거

옵션과 옵션요청DTO간 의존관계 제거

yunjunghun0116 committed:
docs: 2단계 기능 요구 사항 추가

2단계 기능 요구 사항 추가

yunjunghun0116 committed:
refactor: 관리자페이지 form to ajax 로 요청 변경 작업 수행

form to ajax로 요청 변경 작업 수행

yunjunghun0116 committed:
docs: 수정된 기능 요구 사항 반영

수정된 기능 요구 사항 반영

yunjunghun0116 committed:
add: 멤버 엔티티 생성

멤버 엔티티 생성

yunjunghun0116 committed:
refactor: 불필요한 개행을 최소화 한다.

불필요한 개행을 최소화한다.

yunjunghun0116 committed:
feat: MemberRequest 이메일, 패스워드 정규식 추가

이메일, 패스워드 정규식 추가

yunjunghun0116 committed:
refactor: 다양한 형식의 이메일 허용하기 위한 정규식 수정

다양한 형식의 이메일 허용하기 위한 정규식 수정

yunjunghun0116 committed:
feat: 계층간 연결 작업 수행

yunjunghun0116 committed:
feat: 가짜 토큰을 바탕으로 가입을 진행하는 로직 개발

가짜 토큰을 바탕으로 가입을 진행하는 로직 개발

yunjunghun0116 committed:
feat: JWT를 사용해서 ACCESS TOKEN 생성 기능 개발

JWT 라이브러리를 사용해서 액세스토큰 생성

yunjunghun0116 committed:
feat: 중복된 이메일은 DuplicatedEmailException 예외를 발생시킨다.

중복된 이메일은 DuplicatedEmailException 예외를 발생시킨다.

yunjunghun0116 committed:
feat: 유저의 로그인 기능 개발

유저의 로그인 기능 개발

yunjunghun0116 committed:
feat: 잘못된 로그인 요청에 대한 예외처리

잘못된 로그인 요청에 대한 예외처리 - InvalidLoginInfoException 예외 발생

yunjunghun0116 committed:
feat: 이용자 삭제 기능 개발

이용자 삭제 기능 개발

yunjunghun0116 committed:
feat: 인가되지 않은 삭제 요청에 대해서 예외 발생

인가되지 않은 삭제 요청에 대해서 예외 발생

yunjunghun0116 committed:
refactor: AuthUtils로 인증 관련 로직 분리

인증 관련 로직 AuthUtils로 분리

yunjunghun0116 committed:
refactor: secret key의 값을 보호하면서 정적 메서드를 반환할 수 있도록 코드 분리

secret key의 값을 보호하면서 정적 메서드를 반환할 수 있도록 코드 분리

yunjunghun0116 committed:
test: MemberService - 인증 프로세스 테스트 수행

인증 프로세스 테스트 코드 작성 및 수행

yunjunghun0116 committed:
test: 회원 탈퇴하기 로직 테스트 수행

회원 탈퇴하기 로직 테스트 수행

yunjunghun0116 committed:
test: 빈 이름으로 회원가입 요청하기(실패) 테스트 추가

테스트 추가

yunjunghun0116 committed:
test: 허용되지 않는 형식으로 회원가입 요청 테스트 추가

허용되지 않는 형식의 회원가입 요청 테스트 추가

yunjunghun0116 committed:
test: 허용되지 않는 형식의 패스워드로 회원가입 요청하기 테스트 추가

허용되지 않는 패스워드 형식은 회원가입 실패!

yunjunghun0116 committed:
test: 회원가입 후 로그인 및 삭제하는 테스트 추가

회원가입 후 로그인 및 삭제하는 테스트 추가

yunjunghun0116 committed:
remove: 사용하지 않는 함수 제거

사용하지 않는 함수 제거

yunjunghun0116 committed:
docs: 기능요구사항 충족 체크

기능요구사항 충족 체크

yunjunghun0116 committed:
refactor: AuthUtils를 완전 유틸클래스로 변경

완전 유틸클래스로 변경

yunjunghun0116 committed:
refactor: 불필요한 개행 제거 및 누락된 개행 추가

개행 관련 작업

yunjunghun0116 committed:
refactor: 객체지향 프로그래밍 원칙에 따라 인증(토큰)관련 로직을 수행하는 Service계층으로 AuthUtils 이동

객체지향 프로그래밍 원칙에 따라 절차지향적 프로그래밍을 따르는 Util클래스는 상태와 역할을 갖는 Service로 이동

yunjunghun0116 committed:
test: 잘못된 역할로 회원가입 요청 성공실패 테스트 추가

잘못된 역할 타입에 대한 회원가입 요청 테스트 추가

yunjunghun0116 committed:
refactor: 권한에 대한 검증 로직 분리

권한에 대한 검증 로직 분리

yunjunghun0116 committed:
refactor: 스키마 제약조건 추가

스키마 제약조건 추가

yunjunghun0116 committed:
refactor: exist 메소드 limit을 통한 성능 개선

exist 메소드 성능 개선

yunjunghun0116 committed:
refactor: 정의되지 않는 스프링 예외에 대한 처리를 구현해둔 ResponseEntityExceptionHandler 상속

정의되지 않는 스프링 예외에 대한 처리

yunjunghun0116 committed:
refactor: 변수명을 명확하게 수정

변수명을 명확하게 수정

yunjunghun0116 committed:
refactor: 헤더값을 넘겨주면 그에 대한 내부적인 처리는 service계층에서 수행하도록 개선

헤더값만 넘겨주면 그에대한 내부적인 처리는 service 계층에서 수행하도록 개선

yunjunghun0116 committed:
remove: 불필요한 이중 괄호 제거

불필요한 이중괄호 제거

yunjunghun0116 committed:
feat: interceptor를 추가하여 헤더에 대한 검증 로직 컨트롤러에서 분리

컨트롤러에서 헤더 검증 interceptor로 분리작업 수행

yunjunghun0116 committed:
refactor: token 기반으로 로그인된 이용자만 product 데이터에 접근할 수 있도록 코드 리팩토링

로그인된 이용자만 product 데이터에 접근할 수 있도록 코드 리팩토링

yunjunghun0116 committed:
refactor: 로그인 실패 시 401 예외를 발생시킨다.

로그인 실패시 401 예외를 발생

yunjunghun0116 committed:
Merge branch 'yunjunghun0116' into step3

yunjunghun0116 committed:
refactor: Interceptor에서 넘긴 id, role을 RequestAttribute 어노테이션을 통해 읽어오도록 개선

어노테이션을 통해 값을 더 가독성좋게 읽어온다.

yunjunghun0116 committed:
Merge branch 'step3' of https://github.com/yunjunghun0116/spring-gift-wishlist into step3

yunjunghun0116 committed:
remove: merge 과정에서 삭제된 서비스 추가되어 삭제 작업 수행

삭제 작업 수행

yunjunghun0116 committed:
feat: 위시리스트 추가하는 기능 개발

위시리스트 추가 기능 개발

yunjunghun0116 committed:
feat: 존재하지 않는 memberId로 위시리스트 생성시 예외발생

존재하지 않는 memberId로 위시리스트 생성시 예외발생

yunjunghun0116 committed:
refactor: 존재하지 않는 memberId로 접근시 404 -> 401 로 예외 처리 수정

예외 상태코드 수정

yunjunghun0116 committed:
feat: 이용자 별 위시리스트 상품 조회 기능 추가

이용자별 위시리스트 상품 조회 기능 추가

yunjunghun0116 committed:
feat: 위시리스트 상품 수정 기능 개발

위시리스트 상품 수정 기능 개발

yunjunghun0116 committed:
feat: 이미 존재하는 위시리스트상품을 또 추가할경우 기존 위시리스트 상품의 갯수를 추가하는 기능 개발

이미 존재하는 위시리스트 상품을 추가할경우 굳이 새로 만들필요없이 기존 상품의 갯수에 더해주는 방식으로 개발

yunjunghun0116 committed:
feat: 위시리스트 상품이 존재하는지 확인하는 기능 개발

위시리스트 상품이 존재하는지 확인하는 기능 개발

yunjunghun0116 committed:
feat: 위시리스트 상품 수정하는 기능 개발

위시리스트 상품 갯수 수정하는 기능 개발

yunjunghun0116 committed:
docs: README.md에 기능 요구사항 체크

기능 요구사항 체크

asebn1 commented on README.md:
```diff
@@ -24,33 +24,47 @@
+- [X] 이미 위시리스트에 같은 상품이 있을경우 갯수만 추가한다.
+- [X] 위시 리스트에 담긴 상품을 삭제할 수 있다.
+- [X] 상품의 갯수를 0으로 수정할경우 해당 상품은 삭제된다.
+
```
👍 👍

asebn1 commented on src/main/java/gift/config/AuthInterceptor.java:
```diff
@@ -0,0 +1,43 @@
+        var memberRole = authService.getMemberRoleWithToken(token);
+        request.setAttribute("memberRole", memberRole);
+    }
+}
```
config의 `AuthInterceptor`가 `AuthService`를 의존하게 되었어요.
어떻게 수정할 수 있을까요?
- `AuthInterceptor`를 `controller/auth/` 패키지에서 관리하면 어떨까요?
- 로그인, 회원가입을 위한 `AuthController`도 만들어볼까요?

yunjunghun0116 replied:
+ controller/auth 내부에 AuthInterceptor와 AuthController를 두어 인증 관련 컴포넌트의 관리를 더 쉽게 할 수 있도록 개선해보았습니다!
+ 로그인, 회원가입을 위한 AuthController와 이용자의 정보 수정, 회원 탈퇴 등을 위한 MemberController로 구분하는게 관리 측면에서 좋을 것 같아 둘을 분리해보았습니다! 이에 따라 Service 계층도 AuthService와 MemberService로 분리해 관리하도록 개선해보았습니다.

질문에 대한 답변

AuthService를 의존하게 된 이유는 토큰을 복호화 하여 필요한 member_id, member_role 을 가져오는 로직은 비즈니스 로직이라고 생각하고, Service 계층에 위치하도록 했는데 생각해보니 서비스 토큰값이 동일하면 언제든 같은 값을 반환해줘야 하고, 복호화하는 과정은 비즈니스 로직보다는 토큰(인증) 로직에 가깝다 생각하여 AuthInterceptor로 함수를 이전하고, 의존하지 않도록 수정해보았습니다.

yunjunghun0116 replied:
추가질문

현재 AuthInterceptor에서는 로직상 getMemberIdWithToken, getMemberRoleWithToken 과 같은 로직은 private 으로 설정할 수 있지만 테스트하는데 용이하도록 토큰 복호화 후 값을 반환하는 함수는 public으로 만들었는데 괜찮을까요?

asebn1 replied:
테스트만을 위해 private을 public으로 설정하는건 좋지 않다고 생각해요
통합테스트로 진행하거나, 리플렉션을 통해 private을 테스트할 수 있을 것 같아요

asebn1 commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -1,50 +1,43 @@
-        var auth = service.login(loginRequest);
+        var auth = memberService.login(loginRequest);
         return ResponseEntity.ok(auth);
     }
```
로그인, 회원가입은 `MemberController` 보다 `AuthController`에서 관리되는게 더 자연스러울 것 같아요

yunjunghun0116 replied:
로그인, 회원가입 기능은 AuthController에서 관리하고, 이용자의 Name, Email, Password 등의 정보 수정과 같이 이용자의 정보와 관련된 기능을 MemberController에서 관리하는게 더 자연스러울 것 같네요!
감사합니다 :)

asebn1 commented on src/main/java/gift/controller/WishProductController.java:
```diff
@@ -0,0 +1,48 @@
+        this.wishProductService = wishProductService;
+    }
+
+    @PostMapping("/add")
+    public ResponseEntity<Void> addWishProduct(@Valid @RequestBody WishProductAddRequest wishProductAddRequest, @RequestAttribute("memberId") Long memberId) {
+        var wishProduct = wishProductService.addWishProduct(wishProductAddRequest, memberId);
+        return ResponseEntity.created(URI.create("/api/wishes/" + wishProduct.id())).build();
+    }
```
코드 통일을 `var`와 직접적인 타입명 중 `var`로 결정하신 이유가 궁금하네요 ㅎㅎ (정답은 없습니다)

yunjunghun0116 replied:
저는 개인적으로 아래와 같이
1. 함수 내부와 같이 지역 변수로 사용되는 경우에만 var를 사용한다.
2. 전역, 함수의 리턴값, 매개변수 등에는 타입을 명시한다.

위 두가지의 기준을 갖고 사용하게 되면 컴파일러가 2번 기준에 의해 타입을 추론할 수 있으며 타입의 길이가 길어지더라도 변수의 타입 부분은 짧게 가져가면서 타입을 명확히 할 수 있을거 같다는 생각으로 var로 통일하게 되었습니다.

개인적으로 var를 남용하게 되면 오히려 코드를 이해하기 어려울 수 있지만 이 부분을 저는 2번 기준과 함께 
타입의 길이가 짧아진 만큼 변수명에서 ex) savedProduct, registerAuth 등과 같이 조금 더 명확하게 의미를 부여해서 코드를 이해하기 쉬운 코드로 작성하자 라는 생각을 갖고 코드를 작성하고 있습니다!

yunjunghun0116 committed:
test: 관리자의 위시 리스트에 1번 상품을 추가하는 테스트 추가

위시리스트 상품추가 테스트 추가

asebn1 commented on src/main/java/gift/dto/WishProductResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+import gift.model.WishProduct;
+
+public record WishProductResponse(
+        Long id, ProductResponse product, Integer count) {
+    public static WishProductResponse from(WishProduct wishProduct, ProductResponse product) {
+        return new WishProductResponse(wishProduct.getId(), product, wishProduct.getCount());
+    }
+}
```
파라미터로 `WishProduct`를 받고있어요
1. dto에서 도메인 계층을 의존해도 괜찮을까요?
2. service용 dto, controller용 dto를 구분지어보면 어떨까요?

yunjunghun0116 replied:
DTO는 데이터 전송의 목적만을 갖고 있기에 다른 계층을 의존하면 안된다 생각합니다.
때문에 도메인 계층이 아닌 값을 집어 넣도록 수정해서 개선했습니다!

여기서 궁금한게 하나 생겼습니다!
WishProductResponse에 상품id가 아닌, 상품에 대한 정보를 담아 클라이언트에게 반환을 해주고 싶은데 그러기 위해서는 어쩔수 없이 WishProductResponse는 ProductResponse의 정보를 담아야 할 것 같습니다.

이러한 목적을 갖고 같은 DTO 계층간 의존이 필요한 경우에는 의존을 해도 괜찮을까요?
제 경우 선택지는 현재
1. ProductResponse 대신 productId, productPrice, productName 등의 정보까지 입력을 받는다.
2. ProductResponse를 그대로 가져온다.

라고 생각을 합니다. 

각각의 선택지의 단점은
1. ProductResponse의 필드가 추가될 경우, WishProductResponse도 추가해야하는 경우가 발생하는 등 상품의 필드가 추가되거나 변경될 경우 함께 변경해줘야 한다.
2. WishProductResponse가 ProductResponse에 의존하게 된다.

현재 저는 2번 선택지를 택했고, 그 이유는 객체의 이름에서 나타나듯 WishProductResponse는 어떤 상품이 몇개 추가되었는지가 중요하다 생각해서 Product에 의존할 수밖에 없다고 생각했습니다.

이 부분에 대해서 멘토님은 위와 같은 상황에서 둘 중 어떤 선택지를 고를 것 같고(다른 선택지가 있다면 해당 선택지), 그 이유는 무엇인지가 궁금합니다!

+ 2번 질문에 대해서 저는 WishProductService에서 Product에 대한 조회를 수행하고, 클라이언트에게는 해당 상품에 대한 정보중 필요한 정보를 함께 제공한다는 목적에서 구분지었을때 Controller용 DTO와 Service용 DTO에서 각각에 들어가는 정보가 무엇이 달라질까? 고민을 해보았는데 큰 차이를 못느끼겠어서 아직 반영은 하지 않은 상황입니다.

asebn1 replied:
> WishProductResponse는 ProductResponse의 정보를 담아야 할 것 같습니다.
이러한 목적을 갖고 같은 DTO 계층간 의존이 필요한 경우에는 의존을 해도 괜찮을까요?

같은 계층의 dto는 서로 의존해도 문제없다고 생각합니다

> 현재 저는 2번 선택지를 택했고, 그 이유는 객체의 이름에서 나타나듯 WishProductResponse는 어떤 상품이 몇개 추가되었는지가 중요하다 생각해서 Product에 의존할 수밖에 없다고 생각했습니다.
이 부분에 대해서 멘토님은 위와 같은 상황에서 둘 중 어떤 선택지를 고를 것 같고(다른 선택지가 있다면 해당 선택지), 그 이유는 무엇인지가 궁금합니다!

저도 2번이 좋을 것 같아요. 말씀 주신 같은 의견입니다 (ProductResponse의 필드가 추가될 경우, WishProductResponse도 추가해야하는 경우가 발생하는 등 상품의 필드가 추가되거나 변경될 경우 함께 변경해줘야 한다.)

추가로 계층간 의존 방향을 고민해봅시다!
1. Controller 계층에 있는 request dto를 Service 계층에서 의존해도 괜찮을까요?
2. Service 계층에 있는 response dto를 controller 계층에서 의존해도 괜찮을까요?

asebn1 commented on src/main/java/gift/model/WishProduct.java:
```diff
@@ -0,0 +1,42 @@
+        return count;
+    }
+
+    public void updateWishProduct(Integer count) {
+        this.count = count;
+    }
```
count만 수정해주고 있어서 `updateCount`가 명확할 것 같아요

asebn1 commented on src/main/java/gift/repository/MemberJDBCRepository.java:
```diff
@@ -48,11 +48,16 @@ public Member findByEmail(String email) {
+        var sql = "select exists(select 1 from member where id = ? limit 1)";
+        var exists = jdbcTemplate.queryForObject(sql, Boolean.class, id);
+        return exists;
+    }
+
     public boolean existsByEmail(String email) {
-        var sql = "SELECT COUNT(*) FROM member WHERE email = ?";
-        var count = jdbcTemplate.queryForObject(sql, Integer.class, email);
-        if (count > 0) return true;
-        return false;
+        var sql = "select exists(select 1 from member where email = ? limit 1)";
+        var exists = jdbcTemplate.queryForObject(sql, Boolean.class, email);
+        return exists;
```
성능이 개선됐네요 :)

yunjunghun0116 committed:
test: 위시리스트 상품 추가 후 삭제하는 테스트 작성

상품 추가 후 삭제 & 수량 0으로 변경 테스트 작성

asebn1 commented on src/main/resources/schema.sql:
```diff
@@ -1,32 +1,49 @@
+    price     int          not null,
+    image_url varchar(255) not null,
     primary key (id)
 );
 
 create table product_option
 (
     id               bigint auto_increment not null,
-    product_id       bigint,
-    name             varchar(255),
-    additional_price int,
+    product_id       bigint       not null,
+    name             varchar(255) not null,
+    additional_price int          not null,
     primary key (id),
     foreign key (product_id) references product (id)
 );
 
 create table member
 (
     id       bigint auto_increment not null,
-    name     varchar(255),
-    email    varchar(255),
-    password varchar(255),
-    role     varchar(255),
+    name     varchar(255) not null,
+    email    varchar(255) not null unique,
+    password varchar(255) not null,
+    role     varchar(255) not null,
     primary key (id)
 );
```
👍 👍

asebn1 commented on src/test/java/gift/controller/MemberControllerTest.java:
```diff
@@ -107,8 +104,8 @@ void loginFailWithPassword() throws Exception {
 
     @Test
-    @DisplayName("정상적으로 회원가입 후 로그인 요청하기")
-    void registerAndLoginSuccess() throws Exception {
+    @DisplayName("정상적으로 회원가입 후 잘못된 패스워드로 로그인 요청하기")
+    void registerAndLoginFail() throws Exception {
         var register = mockMvc.perform(post("/api/members/register")
                 .contentType(MediaType.APPLICATION_JSON)
                 .content(objectMapper.writeValueAsString(new RegisterRequest("테스트", "test@naver.com", "testPassword", "MEMBER"))));
```
꼼꼼한 테스트코드 좋아요 👍

yunjunghun0116 committed:
test: 이용자 별 보는 위시 리스트가 다른 것을 확인하는 테스트 코드 작성

위시 리스트가 이용자마다 다른 것을 보여주는 테스트 코드 작성

asebn1 reviewed (changes requested):
정훈님 안녕하세요~
빠른 코드반영 너무 좋습니다 👏👏
몇 가지 코멘트 드렸는데 확인해주세요!

yunjunghun0116 committed:
test: 이미 존재하는 위시 리스트의 상품을 추가할 경우 기존 위시리스트의 상품의 갯수가 증가하는 테스트 추가

이미 존재하는 상품을 더 추가할 경우 새로 생성하는게 아닌, 기존 갯수에 추가

yunjunghun0116 committed:
feat: 예기치 못한 서버 예외가 발생할 경우 InternalServerError를 발생

예기치 못한 서버 예외가 발생할 경우 InternalServerError 발생

yunjunghun0116 committed:
refactor: 인증(회원가입, 로그인) 만을 관리하는 AuthController와 이용자의 정보와 관련된 MemberController로 로직 분리

인증로직은 AuthController로 분리, 정보 ex) email 수정, password 수정 등은 MemberController에서 하도록 로직 분리

yunjunghun0116 committed:
refactor: 헤더의 유효성 검사는 AuthService가 아닌 AuthInterceptor에서 검사 하도록 수행

헤더 유효성 검사 로직의 분리

yunjunghun0116 committed:
refactor: WishProductResponse가 도메인 계층을 의존하지 않도록 개선

of를 통해 값만을 입력받도록 개선

yunjunghun0116 committed:
refactor: ResponseDto의 Domain계층 의존도 낮춤

의존성 낮춤 : from -> of 로 메서드 형태를 바꿔버림

yunjunghun0116 committed:
refactor: 토큰 관련 로직을 추상클래스로 분리하고, 필요한 값을 넣어주는 식으로 해서 의존관계를 최소화

의존관계를 최소화 하며 토큰관련 로직은 로직을 제공하는 추상클래스로 구현

yunjunghun0116 committed:
refactor: 추상클래스에서 유틸클래스로 변경

private생성자와 정적메서드 제공을 통한 유틸클래스로 변경

yunjunghun0116 committed:
test: 위시리스트 수량 변경하기 테스트 추가

테스트 추가

yunjunghun0116 committed:
test: 위시리스트 상품 삭제 테스트 추가

위시리스트 상품 삭제 테스트 추가

yunjunghun0116 committed:
test: 단위 수준으로 테스트 잘개 쪼개기

단위 수준으로 테스트 잘개 쪼개기

yunjunghun0116 committed:
test: 테스트 단위 수준으로 잘개 쪼개기

테스트 단위 수준으로 잘개 쪼개기

yunjunghun0116 committed:
refactor: 유틸 클래스의 기능을 AuthInterceptor로 옮기고 유틸클래스와의 결합도를 낮춤

유틸 클래스의 기능을 옮기고 유틸클래스와의 결합도를 낮춤

yunjunghun0116 committed:
refactor: 인증 로직과 멤버 로직의 분리(서비스계층)

인증 로직과 멤버 로직의 분리

yunjunghun0116 committed:
move: auth와 member로 나누기 위해 controller, service, test 모두 auth 폴더 생성후 로직 분리

yunjunghun0116 committed:
test: 인증 관련 테스트와 로직(삭제) 로직 테스트 분리

인증관련 테스트와 로직 관련 테스트 분리

yunjunghun0116 committed:
docs: README.md 수정 및 요구사항 정리

README.md 수정 및 요구사항 정리

asebn1 reviewed (approved):
정훈님 안녕하세요~ 이번에도 빠르게 구현해주셨군요 👏👏
잘 구현해주셔서 3단계는 이만 머지하겠습니다 :)
